server/src/main/java/org/elasticsearch/search/retriever/RankDocsRetrieverBuilder.java

```diff
@@ -124,19 +130,39 @@ public void extractToSearchSourceBuilder(SearchSourceBuilder searchSourceBuilder
             boolQuery.filter(preFilterQueryBuilder);
         }
         searchSourceBuilder.query(boolQuery);
+        searchSourceBuilder.rankBuilder(
+            new RankDocsRankBuilder(rankWindowSize, sources.stream().map(RetrieverBuilder::retrieverName).toList())
```

Comment:
I see your point. Tbh conceptually I believe that, similarly to how a `Weight` explains how a hit was generated, a retriever/ranker should also be able to explain the operation that it applies and how it combines/ranks different results (aside from having just the `RankDoc#explain`). Maybe we could use something besides the `RankBuilder` for that (as it seems that this is its only purpose now) but I think that it should probably be something global (i.e. know the retriever and the rank function) and available throughout the request.  

However, as you mentioned, nested compound retrievers will be more of a challenge, as we'd need access to multiple different retrievers and explanations. E.g. considering the 
```
rrf: 
       - semantic_reranker:
           - rrf
               - knn
               - standard
      - rrf:
           - standard
           - standard
```
When we're done with the query rewriting phase, we'd rewrite the top-level rrf query to a `RankDocs` with `RRFRankDocs` associated objects. All info on original parsed retriever, and all the nested ones and how the scores were computed will have been "lost" for the context of explaining.

Reply:
> Maybe we could use something besides the RankBuilder for that (as it seems that this is its only purpose now) but I think that it should probably be something global (i.e. know the retriever and the rank function) and available throughout the request.

Perhaps we could update the API to introduce `RankDocs<T extends RankDoc>` and provide a way to package explain information for a set of results. I do believe the rank API should be reserved exclusively for the global re-ranker.

Although, I wonder if we could improve the explain functionality in a follow-up. For now, we could include the explanation for the final level in the response. It's straightforward to run the sub-requests individually, just like normal retrievers, to get additional information from them.

Reply:
+1 on revisiting this in a follow-up as it seems a bit of a complex issue. I'd suggest however to keep the `RankBuilder#explainHit` for now, as the `RRFRankBuilder` (which is used by `rank` and `sub_searches`) still relies on it and would increase the scope of this PR significantly.  WDYT?

Reply:
+1 on revisiting this in a follow-up as it seems a bit of a complex issue. I'd suggest however to keep the RankBuilder#explainHit for now if you're ok with this, as the RRFRankBuilder (which is used by rank and sub_searches) still relies on it and would increase the scope of this PR significantly. WDYT?

Reply:
I'd prefer if we can avoid adding a new rank builder implementation. Can we start simple and rely on the fact that  the `RankDocsQuery` has already an explain? We can add the explain by query name in a follow up.

Reply:
`RankDocsQuery#explain` currently relies on just the provided `RankDoc` results, which hold little information of the underlying retrievers. For RRF we do have scores and positions, but that's about it. We can omit the ranker if you prefer, but I think that this would remove all underlying information on the child retrievers.  So the output would change from something like this -> 
```
     "_explanation": {
          "value": 0.01980198,
          "description": "rrf score: [0.01980198] computed for initial ranks [1, 1] with rankConstant: [100] as sum of [1 / (rank + rankConstant)] for each query",
          "details": [
            {
              "value": 1,
              "description": "rrf score: [0.00990099], for rank [1] in query at index [1] computed as [1 / (1 + 100)], for matching query with score [my-standard-2]",
              "details": [
                {
                  "value": 0.9808291,
                  "description": "weight(text:some_term in 1) [PerFieldSimilarity], result of:",
                  "details": [
```
to
```
     "_explanation": {
          "value": 0.01980198,
          "description": "rrf score: [0.01980198] computed for initial ranks [1, 1] with rankConstant: [100] as sum of [1 / (rank + rankConstant)] for each query",
          "details": [
            {
              "value": 1,
              "description": "rrf score: [0.00990099], for rank [1] in query at index [1] computed as [1 / (1 + 100)], for matching query with score 0.19",
              "details": [
               ]
          },
          {    
              "value": 1,
              "description": "rrf score: [0.00990099], for rank [1] in query at index [1] computed as [1 / (1 + 100)], for          matching query with score 0.5",
             "details": []
}
```

i.e. the `details` under each rrf entry will be null and we would lose information on the underlying
```
"description": "weight(text:some_term in 1) [PerFieldSimilarity], result of:"
```

Reply:
What about folding the original query into the RankDocsQueryBuilder/RankDocsQuery? Then we can make the RankDocsQuery executes the scorer depending on whether we need total hits and/or original scores (for explain purposes).
This way all the explain can be in a single place with the right tree. That would be better than having to link queries through their names in the explain?
In fact this is aligned with the original intent, a retriever that rewrites to a query should be explainable through it.

Reply:
I like the idea tbh (it was also briefly discussed in our team's meeting but it wasn't materialzied). Will look into this and come back soon!

Reply:
Added a rough take on this in https://github.com/elastic/elasticsearch/commit/06ec5387029b0f101c8752915420962210cf82b7.  As discussed, we can maybe also look into implementing all `topDocs` functionality from within the `RankDocsQuery`.